[CCI] Create advanced index actions guide (#147)
opensearch-project/opensearch-rs: guides/advanced_index_actions.md

# Advanced Index Actions

> In this guide, we will look at some advanced index actions that are not covered in the [Index Lifecycle](index_lifecycle.md) guide.

Reviewer: That file doesn't exist yet, so the link checker breaks. You can combine all the guides in 1 PR or make them in order they link to each-other.

Author: @dblock  Can we wait for the index lifecycle to merge first and then re-run the link checker?

## Setup

Let's create a client instance, and an index named `movies`:

```rust

> // Create a client to make API calls to OpenSearch running on http://localhost:9200.
> let client = OpenSearch::default();
> // Alternatively, you can create a client to make API calls against OpenSearch running on a specific url::Url.
> let url = Url::parse("https://example.com")?;
> let transport = TransportBuilder::new(SingleNodeConnectionPool::new(url)).cert_validation(CertificateValidation::None).build()?;
> let client = OpenSearch::new(transport);

Reviewer 2: Let's make this consistent with the other guides:
```suggestion
// Create a client to make API calls to OpenSearch running on https://localhost:9200.
let url = Url::parse("https://localhost:9200")?;
let credentials = Credentials::Basic("admin".into(), "admin".into());
let transport = TransportBuilder::new(SingleNodeConnectionPool::new(url))
      .cert_validation(CertificateValidation::None)
      .auth(credentials)
      .build()?;
let client = OpenSearch::new(transport);
```

client.indices().create(IndicesCreateParts::Index("movies")).send().await?;
```

## API Actions

### Clear index cache

You can clear the cache of an index or indices by using the `indices.clear_cache` API action. The following example clears the cache of the `movies` index:

```rust
client.indices().clear_cache(IndicesClearCacheParts::Index(&["movies"])).send().await?;
```

By default, the `indices.clear_cache` API action clears all types of cache. To clear specific types of cache pass the the `query`, `fielddata`, or `request` parameter to the API action:

```rust

> client.indices().clear_cache(IndicesClearCacheParts::Index(&["movies"])).query(true).send().await?;
>     client.indices().clear_cache(IndicesClearCacheParts::Index(&["movies"])).fielddata(true).request(true).send().await?;

Reviewer 2: The indentation is a little weird here, let's fix it like so:
```suggestion
client.indices().clear_cache(IndicesClearCacheParts::Index(&["movies"])).query(true).send().await?;
client.indices().clear_cache(IndicesClearCacheParts::Index(&["movies"])).fielddata(true).request(true).send().await?;
```

```

### Force merge index

You can force merge an index or indices to reduce the number of segments in the index. This can be useful if you have a large number of small segments in the index. Merging segments reduces the memory footprint of the index. Do note that this action is resource intensive and it is only recommended for read-only indices. The following example force merges the `movies` index:

```rust
client.indices().forcemerge(IndicesForcemergeParts::Index(&["movies"])).send().await?;
```

### Clone index

You can clone an index to create a new index with the same mappings, data, and MOST of the settings. The source index must be in read-only state for cloning. The following example blocks write operations from `movies` index, clones the said index to create a new index named `movies_clone`, then re-enables write:

```rust
client.indices().add_block(IndicesAddBlockParts::IndexBlock(&["movies"],"write")).send().await?;
client.indices().clone(IndicesCloneParts::IndexTarget("movies","movies_clone")).send().await?;
client.indices().put_settings(IndicesPutSettingsParts::Index(&["movies"]))
    .body(json!({

>     "index": {
>         "blocks": {
>             "write": false
>         }
>     }

Reviewer 2: ```suggestion
        "index": {
            "blocks": {
                "write": false
            }
        }
```

    })).send().await?;
```

### Split index

> You can split an index into another index with more primary shards. The source index must be in read-only state for splitting. The following example create the read-only `books` index with 30 routing shards and 5 shards (which is divisible by 30), splits index into `bigger_books` with 10 shards (which is also divisible by 30), then re-enables write:

Reviewer 2: ```suggestion
You can split an index into another index with more primary shards. The source index must be in read-only state for splitting. The following example creates the read-only `books` index with 30 routing shards and 5 shards (which 30 is divisible by), splits the index into `bigger_books` with 10 shards (which 30 is also divisible by), then re-enables writes:
```

```rust
client.indices().create(IndicesCreateParts::Index("books"))
    .body(json!({

>     "settings": {
>         "index": {
>             "number_of_shards": 5,
>             "number_of_routing_shards": 30,
>             "blocks": {
>                 "write": true
>             }
>         }
>     }

Reviewer 2: ```suggestion
        "settings": {
            "index": {
                "number_of_shards": 5,
                "number_of_routing_shards": 30,
                "blocks": {
                    "write": true
                }
            }
        }
```
